src/sage/groups/additive_abelian/additive_abelian_wrapper.py

```diff
@@ -566,3 +616,188 @@ def _rec(j, k, c):
         return x
 
     return _rec(0, max(vals), b)
+
+
+def _expand_basis_pgroup(p, alphas, vals, beta, h, rel):
+    r"""
+    Given a basis of a `p`-subgroup of a finite abelian group
+    and an element lying outside the subgroup, extend the basis
+    to the subgroup spanned jointly by the original subgroup and
+    the new element.
+    Used as a subroutine in :func:`basis_from_generators`.
+
+    This function modifies ``alphas`` and ``vals`` in place.
+
+    ALGORITHM: [Suth2007]_, Algorithm 9.2
+
+    EXAMPLES::
+
+        sage: from sage.groups.additive_abelian.additive_abelian_wrapper import _expand_basis_pgroup
+        sage: A = AdditiveAbelianGroup([9,3])
+        sage: alphas = [A((5,2))]
+        sage: beta = A((1,0))
+        sage: vals = [2]
+        sage: rel = next([ZZ(r),ZZ(s)] for s in range(9) for r in range(9) if s > 1 and not r*alphas[0] + s*beta)
+        sage: _expand_basis_pgroup(3, alphas, vals, beta, 2, rel)
+        sage: alphas
+        [(5, 2), (6, 2)]
+        sage: vals
+        [2, 1]
+        sage: len({i*alphas[0] + j*alphas[1] for i in range(3^2) for j in range(3^1)})
+        27
+    """
+    k = len(rel)
+    assert isinstance(alphas, list) and isinstance(vals, list)
+    assert len(alphas) == len(vals) == k-1
+    assert all(r >= 0 for r in rel)
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+#    assert all(a.order() == p**v for a,v in zip(alphas,vals))
+
+    # step 1
+    for i in range(k-1):
+        if not rel[i]:
+            continue
+        q = rel[i].p_primary_part(p)
+        alphas[i] *= rel[i]//q
+        rel[i] = q
+    min_r = min(filter(bool, rel))
+    val_rlast = rel[-1].valuation(p)
+    assert rel[-1] == p**val_rlast
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+
+    # step 2
+    if rel[-1] == min_r:
+        for i in range(k-1):
+            beta += alphas[i] * (rel[i]//rel[-1])
+        alphas.append(beta)
+        vals.append(val_rlast)
+#        assert alphas[-1].order() == p**vals[-1]
+        return
+
+    # step 3
+    j = next(j for j,r in enumerate(rel) if r == min_r)
+    alphas[j] = sum(a * (r//rel[j]) for a,r in zip(alphas+[beta], rel))
+
+    # step 4
+    if not alphas[j]:
+        del alphas[j], vals[j]
+        if not alphas:
+            alphas.append(beta)
+            vals.append(val_rlast)
+#            assert alphas[-1].order() == p**vals[-1]
+            return
+
+    # step 5
+    beta_q = beta
+    for v in range(1, h):
+        beta_q *= p
+        try:
+            e = _discrete_log_pgroup(p, vals, alphas, -beta_q)
+        except TypeError:
+            continue
+        else:
+            # step 6
+            _expand_basis_pgroup(p, alphas, vals, beta, h, list(e) + [p**v])
+#            assert alphas[-1].order() == p**vals[-1]
+            return
```

Comment:
This was indeed unnecessarily complicated, does it look better without the inner `else`?

Reply:
There’s that (which I think does make it a bit more clear). However, there is also the fact you are using a `try-except` block to basically find one particular object rather than an `if` statement with a more localized test. It has a whiff of a code smell to me, that something should be refactored, but it isn’t anything that needs to be done here unless you can see an “easy” solution.

Reply:
In general, I think all the `.discrete_log()` methods should return `None` instead of throwing exceptions when the logarithm doesn't exist — but I would leave that for another ~~ticket~~ issue since it's a breaking change. The `TypeError` here was inherited from the original implementation of `AdditiveAbelianGroupWrapper._discrete_log()`.

Reply:
I concur. I also don't immediately see a way to extract a more simple check (although not really my area of expertise).